묵찌빠 게임 [STEP 2] LJ, 준호

junho15 opened the pull request:
안녕하세요 🍞브래드(@bradheo65)
LJ(@lj-7-77), 준호(@junho15) 입니다. 
STEP 2 잘부탁드립니다.

### 고민했던 점
- 가위바위보 게임의 startGame 메서드명은 게임을 시작한다는 의미인데 반환값이 있다는 것이 어색했습니다.
- 해결방법 ➡️ winner 프로퍼티를 추가해서 가위바위보 게임이 끝나면 승자를 winner 프로퍼티에 설정하여 묵찌빠게임에서 쓸 수 있도록 했습니다.
- 수정 전
```swift
struct RockPaperScissorsGame {
    func startGame() -> Participant? {
        
```
- 수정 후
```swift
struct RockPaperScissorsGame {
    var winner: Participant?
    
    mutating func startGame() {
```
---
- 묵찌빠 게임에서 먼저 가위바위보 게임이 끝나고 승리자가 결정된 후 묵찌빠 게임을 진행하는 부분은 doMukChiBaGameRoutine() 메서드로 분리했습니다.
~~~swift
    mutating func startGame() {
        var rockPaperScissorsGame: RockPaperScissorsGame = RockPaperScissorsGame()
        rockPaperScissorsGame.startGame()
        guard let turnOwner: Participant = rockPaperScissorsGame.winner else {
            return
        }
        self.turnOwner = turnOwner
        
        doMukChiBaGameRoutine()
    }
    
    mutating private func doMukChiBaGameRoutine() {
        var isTurnOwnerWin: Bool
        repeat {
            let selectedUserMenu: Int = getSelectedUserMenu()
            guard let userHand: MukChiBa = MukChiBa.init(rawValue: selectedUserMenu) else {
                return
            }
            guard let gameResult: GameResult = compareComputerHand(with: userHand) else {
                return
            }
            isTurnOwnerWin = isTurnOwnerMukChiBaGameWin(gameResult: gameResult)
            printMukChiBaGameResult(isTurnOwnerWin)
        } while !isTurnOwnerWin
    }
~~~
---
- repeat 블록의 코드를 최초 1회 실행한 후, while 조건이 성립하는 지를 확인하기 위해서 repeat-while 문을 사용했습니다. 
~~~swift
    mutating private func doMukChiBaGameRoutine() {
        var isTurnOwnerWin: Bool
        repeat {
            let selectedUserMenu: Int = getSelectedUserMenu()
            guard let userHand: MukChiBa = MukChiBa.init(rawValue: selectedUserMenu) else {
                return
            }
            guard let gameResult: GameResult = compareComputerHand(with: userHand) else {
                return
            }
            isTurnOwnerWin = isTurnOwnerMukChiBaGameWin(gameResult: gameResult)
            printMukChiBaGameResult(isTurnOwnerWin)
        } while !isTurnOwnerWin
    }
~~~

---

### 조언을 얻고 싶은 부분
- 가위바위보게임과 묵찌빠게임과정에서 예로 가위vs바위 경우 바위가 승리하는 승패결정규칙은 동일합니다. 요구사항 명세서에는 가위(1),묵찌빠 중 찌(2)로 부여된 숫자값이 달라 따로따로 구현하였습니다. 그러나 순서도를 보면 중복인 부분이 보입니다.
- 만약 가위(1)와 찌(2)가 같은숫자로 부여된다면,
- 그리고 가위바위보 승패결정 메서드에 private이 없다면,
- 묵찌빠게임에서 가위바위보게임에 승패결정 메서드에 접근하여 가져다 쓸 수 있을까요?

---

- 메서드 내부에서 다른 메서드를 호출할 때 마다 turnOwner를 매개변수로 전달하는 방법과 프로퍼티를 선언해서 여러 메서드에서 사용하는 방법 중에 프로퍼티를 사용하는 방법을 선택했습니다.
- turnOwner 프로퍼티는 인스턴스를 처음 생성할 때 초깃값을 할당해 주지 않아도 되도록 옵셔널 타입으로 했습니다. 
- 매번 매개변수로 전달해 줄 필요 없이, 필요한 메서드에서 turnOwner 프로퍼티의 값을 쓰거나 읽으니 편하기는 한데 이렇게 하는 게 좋은 방법인지 조언을 얻고 싶습니다. 
~~~swift
    // 8~9줄
    struct MukChiBaGame {
        private var turnOwner: Participant?
           
    // 121~127 줄
    mutating private func toggleTurnOwner() {
        if turnOwner == .user {
            turnOwner = .computer
        } else {
            turnOwner = .user
        }
    }
        
    // 129~139줄
    private func printMukChiBaGameResult(_ isTurnOwnerWin: Bool) {
        guard let turnOwnerName = turnOwner?.name else {
            return
        }
        switch isTurnOwnerWin {
        case true:
            print("\(turnOwnerName)의 승리!")
        case false:
            print("\(turnOwnerName)의 턴입니다.")
        }
    }
~~~



--- 
### 순서도
<img src="https://cdn.discordapp.com/attachments/1011075486962679979/1012169701989810186/STEP2-STEP_2.drawio.png">

junho15 committed:
feat: 묵찌빠 열거형 추가

limjy0 committed:
feat: 참가자 열거형 추가

junho15 committed:
feat: 묵찌빠 구조체 추가, 참가자 열거형 수정

- 묵찌빠 구조체 추가
- 묵찌빠 게임 메뉴 출력 메서드 추가
- 참가자 열거형 이름 첫글자 대문자로 수정
- 참가자 열거형에 프로퍼티 name 추가

limjy0 committed:
refactor: 가위바위보게임 종료시 승리자 반환하도록 기능 수정

junho15 committed:
Merge branch 'step1' into junho-feature-1

# Conflicts:
#	RockPaperScissors/RockPaperScissors.xcodeproj/project.pbxproj
#	RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift

limjy0 committed:
feat: 사용자입력값 관련 메서드 추가

- getSelectedUserMenu 입력값검증 후 잘못입력 시 에러메세지 출력, 정상입력 시 입력값 반환
- userInput 입력값 받는 메서드
- isCorrectUserMenu 입력값이 0~3범위인지 여부확인 메서드

junho15 committed:
feat: 승패를 결정하는 메서드 추가

- 승패를 결정하는 메서드 compareComputerHand 추가
- 컴퓨터의 패를 랜덤하게 정하는 메서드 getComputerHand 추가

limjy0 committed:
feat: 묵찌빠게임 구조체 프로퍼티 추가, 턴오너 관련 메서드 추가

- 메서드에서 사용하는 turnOwner 파라미터를 turnOwner 프로퍼티로 선언하여 사용
- 턴오너를 교체하는 메서드 toggleTurnOwner
- 턴오너가 이겼는지 판단하는 메서드 isTurnOwnerWin

junho15 committed:
feat: 묵찌빠 게임 루틴 실행 메서드 추가

- 묵찌빠 게임 루틴 실행하는 doMukChiBaGameRoutine 메서드 추가
- isTurnOwnerMukChiBaGameWin로 메서드 이름 수정

limjy0 committed:
feat: 누구의 턴인지 알리는 출력문 추가

- doMukChiBaGameRoutine 메서드 switch문에서 패가 다른경우 턴을 쥐고있는 쪽이 누구인지 출력

junho15 committed:
refactor: 가위바위보 게임 startGame 메서드에서 if문을 switch 문으로 리팩토링

limjy0 committed:
refactor: 메서드 순서, 분리 리팩토링

- 메서드 배치 호출순으로 변경
- doMukChiBaGameRoutine메서드 결과출력부분 printMukChiBaGameResult메서드로 분리
- doMukChiBaGameRoutine메서드 guard문 &&조건식을 guard문 두 개로 분리

junho15 committed:
refactor: startGame 메서드 수정

- 가위바위보 게임 startGame 메서드에서 Participant를 직접 반환하지 않고 winner 프로퍼티 값을 바꿔주도록 수정
- 묵지빠 게임 startGame 메서드에서 가위바위보 게임 진행 후 winner 프로퍼티 값을 확인해서 turnOwner를 설정하도록 수정.

bradheo65 commented on RockPaperScissors/RockPaperScissors/MukChiBaGame.swift:
```diff
@@ -0,0 +1,140 @@
+            }
+            isTurnOwnerWin = isTurnOwnerMukChiBaGameWin(gameResult: gameResult)
+            printMukChiBaGameResult(isTurnOwnerWin)
+        } while !isTurnOwnerWin
```
`repeat while`로 조건문 구현해 주셨네요 정말 좋습니다!!! 
개인적인 의견이지만 `!` 로 한눈에 보기 어려울 수 있을 것 같아요. `isTurnOwnerWin = false`로 명시해줘도 가독성이 좋아질 것 같습니다!!

bradheo65 commented on RockPaperScissors/RockPaperScissors/MukChiBaGame.swift:
```diff
@@ -0,0 +1,140 @@
+            return getSelectedUserMenu()
+        }
+        
+        if isCorrectUserMenu(selectedUserMenu) {
+            return selectedUserMenu
+        } else {
+            print("잘못된 입력입니다. 다시 시도해주세요.")
+            turnOwner = .computer
+            return getSelectedUserMenu()
+        }
```
```swift
        if isCorrectUserMenu(selectedUserMenu) {
            return selectedUserMenu
        }
        print("잘못된 입력입니다. 다시 시도해주세요.")
        turnOwner = .computer
        return getSelectedUserMenu()
```
로 변경 하면 어떻게 될까요!?
추가적으로 이 부분은 메소드로 따로 관리해줘도 괜찮을 것 같습니다!!

팁을 드리자면 이런 방법도 있어요
![image](https://user-images.githubusercontent.com/45350356/186712242-230b2a38-7432-411a-9b44-0b0a4b49f740.png)

jylim72 replied:
아 저런 방법도 있었네요! 꿀팁🍯 감사합니다👍

bradheo65 commented on RockPaperScissors/RockPaperScissors/MukChiBaGame.swift:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    mutating private func toggleTurnOwner() {
+        if turnOwner == .user {
+            turnOwner = .computer
+        } else {
+            turnOwner = .user
+        }
```
```swift
       if turnOwner == .user {
            turnOwner = .computer
            return
        }
        turnOwner = .user
```
이 부분도 위의 형식으로 바꾸면 어떨까요??

jylim72 replied:
생각해보니 굳이 else문을 써 줄 이유가 없네요! 코드 한 줄도 왜 라는 이유를 생각하는 습관을 들이겠습니다🤓

bradheo65 commented on RockPaperScissors/RockPaperScissors/MukChiBaGame.swift:
```diff
@@ -0,0 +1,140 @@
+        rockPaperScissorsGame.startGame()
+        guard let turnOwner: Participant = rockPaperScissorsGame.winner else {
+            return
+        }
+        self.turnOwner = turnOwner
```
`guard let` 끝나는 지점에 붙이는 부분도 있고 개행 해주시는 부분도 있는 것 같아요!! 
통일해주시면 좋을 것 같습니다!!

jylim72 replied:
`guard let` 끝나는 지점에 붙이는 것으로 코딩컨벤션을 정했습니다.
추가로 `return`문이 속한 코드블럭이 여러줄이면 `return`문 윗 한줄 개행하고,
코드블럭에 `return`문 한줄만 있다면 윗줄과 붙여쓰는 컨벤션을 정하였습니다.

bradheo65 replied:
넵 따로 정해주신 컨벤션만 있다면 상관없을 것 같습니다!! 코드 형식엔 정답이 없으니까요!!

bradheo65 commented on RockPaperScissors/RockPaperScissors/RockPaperScissorsGame.swift:
```diff
@@ -6,15 +6,24 @@
-            if gameResult == .draw {
-                startGame()
-            }
+        guard let userHand: RockPaperScissors = RockPaperScissors.init(rawValue: selectedUserMenu),
+           let gameResult: GameResult = compareComputerHand(with: userHand) else {
+            return
+        }
```
개행 맞춰주세요!!

bradheo65 commented on RockPaperScissors/RockPaperScissors/MukChiBaGame.swift:
```diff
@@ -0,0 +1,140 @@
+    mutating private func getSelectedUserMenu() -> Int {
+        printUserMenu()
+        guard let selectedUserMenu: Int = userInput() else {
+            print("잘못된 입력입니다. 다시 시도해주세요.")
```
print 하는 메소드가 `printUserMenu()` 있는데 여기도 메소드로 관리를 해주시면 좋을 것 같습니다.

jylim72 replied:
넵 출력해주는 동작은 같은데 방법이 통일되지 않았었습니다. 말씀주신대로 아래와 같이 수정하였습니다. 🤓
```swift
    private func printWrongInputMessage() {
        print("잘못된 입력입니다. 다시 시도해주세요.")
    }
```

bradheo65 replied:
저도 아직 부족한 부분이지만 코드의 일정한 형식을 생각해서 코드 작성 하면 좋을 것 같습니다!!

bradheo65 commented on RockPaperScissors/RockPaperScissors/MukChiBaGame.swift:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    private func printUserMenu() {
+        guard let turnOwnerName = turnOwner?.name else {
+            return
+        }
```
네이밍이 어색해 보여요!!

jylim72 replied:
turnOwnerName 에 turnOwner의 name을 넣는코드가 같은내용을 똑같게 읽히는 상수에 넣는것으로 해석이 되어 저희도 다시 보니 어색하네요😅 nameToPrint로 수정하였습니다.
```swift
 guard let nameToPrint = turnOwner?.name else {
     return
}
```

bradheo65 commented on RockPaperScissors/RockPaperScissors/MukChiBaGame.swift:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    private func userInput() -> Int? {
+        let userInput: Int? = Int(readLine() ?? "")
+        return userInput
```
` ?? ` 로 옵셔널 처리를 해주신 이유가 있을까요~?

jylim72 replied:
간단하게 unwrapping 하기 위하여 사용하였습니다. 
`if let` 이나 `guard let`을 사용하면 코드가 더 길어져서 `??`로 처리를 해주었습니다.
사용자입력값을 받는 readLine() 값을 받을 다른 좋은 방법이 있을까요? 🤔

bradheo65 replied:
제가 생각한 방향은 이렇게 코드를 표현해보았어요!!

```swift
guard let userInput = readLine() else {
 return
}
let userNumberInput = userInput
return user Input
```
지금 프로젝트는 아니지만 `readLine()`에 여러가지 메소드가 붙는 경우가 생길 수도 있어요!!
예를 들어 `readLine().components()` 등등 
그 상황을 대비해서 미리 분리해두면 좋지 않을까 생각 되서 말씀 드린겁니다!!

jylim72 replied:
아하 네 말씀 듣고 생각해보니 그러네요!! 미리 분리해두면 좀 더 세분화된 구현을 할 때에도 좋을 것 같습니다!! 조언 감사합니다👍

bradheo65 commented on RockPaperScissors/RockPaperScissors/MukChiBaGame.swift:
```diff
@@ -0,0 +1,140 @@
+        doMukChiBaGameRoutine()
+    }
+    
+    mutating private func doMukChiBaGameRoutine() {
```
`startMukChiBaGameRoutine`로 변경되면 어떨까요~?

jylim72 replied:
startGame 메서드도 있어 말씀주신대로 통일하니 더 보기 좋습니다.👍

bradheo65 reviewed (commented):
STEP2도 정말 고생많으셨습니다 LJ, 준호!!💯💯
코드가 정말 좋은 것 같아요.. 제가 오히려 많이 배운 것 같네요 ㅎㅎ..🙇🏻‍♂️
코드들에 대한 리뷰를 작성해 보았는데요!! 로직적인 것 보단 네이밍, 컨벤션, 메소드 분리에 관해서 드려 보았습니다. 

타입 분리, 메소드 분리, repeat while 활용도 굿입니다!!👍

### 답변 

#### A1. 

예를 들어 묵찌바게임 가위바위보게임 각 객체가 있다면 게임 순서가 가위바위보 -> 묵찌바로 진행이 될텐데 해당 가위바위보의 승자에 대한 값을 묵찌바 객체에 있는 저장 프로퍼티에 저장해주는 식으로 가져다 쓸 수 있을 것 같습니다

추후에도 View 간 데이터 전달함에 있어서 공부를 할 수 있는 기회가 있을 겁니다!!

아래 내용 참고해주세요!
[데이터 전달](https://i-colours-u.tistory.com/6)

---

#### A2. 

넵 제가 생각했을 때의 느낌을 의견을 드리자면 옵셔널 타입으로 프로퍼티로 만들어준 건 정말 좋은 접근 인것 같습니다. 일단 게임 턴에 대한 건 가위바위보 게임이 끝나기 전까진 모르는 상태이기에 옵셔널 이라는 뜻과 잘 어울리는 것 같고 사용되는 메소드 안에서 옵셔널 바인딩과 필요한 로직을 해주신것에 대해서는 코드 읽기 편했습니다.👍👍

#### 추가 코멘트

추가적으로 순서도 구현을 구체적으로 잘해주셨는데요. 

개인적인 생각으로는 순서도가 한눈에 보기 좋게 만들어 보는 것도 추천드립니다. 해당 비즈니스 로직에 대해서는 필요한 부분이 맞지만 이 프로그램이 어떤 플로우로 진행이 될 것인지 한 눈에 파악하면 좋지 않을까 생각됩니다!!

limjy0 committed:
refactor: 리뷰 수정사항 반영

- doMukChiBaGameRoutine메서드명 startMukChiBaGameRoutine 로 변경
- repeat while조건문 수정
- printWrongInputMessage 메서드 추가
- printUserMenu메서드에 상수명 nameToPrint로 변경
- return문 개행 코딩컨벤션 추가
- getCorrectUserMenu 메서드 추가
- toggleTurnOwner 메서드 else문 제거

junho15 committed:
docs: README 파일 작성

limjy0 committed:
refactor: README파일 flow-chart, 순서도 리팩토링

limjy0 committed:
Merge branch 'lj-feature-1' into step2

refactor: REAMD파일 flow chart, 실행화면 리팩토링

limjy0 commented:
README 작성 시 놓칠 수 있는 부분들까지 리뷰해주시고, 네이밍과 같은 중요한 팁도 많이 전수해주셔서 감사합니다. 코드가독성과 완성도가 많이 높아졌습니다. 브래드🍞 시간내주셔서 감사해요. 고생 많으셨습니다.👍👍👍
